Machine ID: Refactor `tbot` CLI

timothyb89 opened the pull request:
This significantly refactors the tbot CLI. In addition to a major overhaul of our CLI config handling, it also exposes a number of new subcommands for starting more than just an identity output via pure CLI.

### New command examples

```bash
# old commands are now under the (default) `tbot start legacy`, so the `tctl bots add` example still works
tbot start --destination-dir=./tbot-user \                                                                                                                                                                    
  --token=<snip> \
  --proxy-server=example.com:443 \
  --join-method=token --data-dir=./tbot-data

# start in identity mode (roughly equivalent to legacy, but with new-style flags)
$ tbot start identity --destination=./tbot-user --storage=./tbot-data --proxy-server=example.com:443 --join-method=token

# most commands have common aliases to improve memorability
$ tbot start ssh --destination=./tbot-user --storage=./tbot-data --proxy-server=example.com:443 --join-method=token

# start in Kubernetes mode
$ tbot start kubernetes --destination ./tbot-user --storage=./tbot-data --proxy-server=example.com:443 --join-method=token --kubernetes-cluster kind-demo

# start a database tunnel
$ tbot start db-tunnel --listen tcp://0.0.0.0:1234 --storage=./tbot-data --proxy-server=example.com:443 --join-method=token --service postgres-demo --username demo --database demo

# start the SPIFFE SVID output
$ tbot start spiffe-x509-svid  --destination=./tbot-user --storage=./tbot-data --proxy-server=example.com:443 --join-method=token --svid-path /foo
```

These commands and flags all now have 1:1 parity with `tbot configure`. That means you can easily convert any of these example commands to a config file by replacing `tbot start` with `tbot configure`:

```bash
# try running a bot manually to test flags
$ tbot start spiffe-x509-svid --destination ./tbot-user --storage=./tbot-data --proxy-server=example.com:443 --join-method=token --svid-path=/foo

# once satisfied, easily convert to a config file
$ tbot configure spiffe-x509-svid --destination ./tbot-user --storage=./tbot-data --proxy-server=example.com:443 --join-method=token --svid-path=/foo > tbot.yaml

# start from the config file
$ tbot start -c tbot.yaml
```

The `--destination` and `--storage` flags now also universally accept URIs, so `memory://` and `kubernetes-secret://` should work as expected.

### Goals

Some (re-)design goals per #44509 include:
1. Exposing various tbot output types and services through pure CLI
2. Cleanly cutting over to new terminology for all new subcommands (`--storage` instead of `--data-dir`, `--destination` that accepts URIs instead of `--destination-dir`, etc)
3. Maintaining support for legacy syntax

While iterating, I also prioritized:
1. Sharing as much code as possible between each subcommand, as well as the `tbot start` and `tbot configure` variants of each. An earlier iteration unrolled all of this, and bugs very rapidly resulted from drift between copy and pasted code blocks, not to mention the spectacular line of code count.
2. Moving away from the effectively global namespace of the `CLIConf` struct. This struct has become unmanageable and led to conflicting usages and definitions of fields. Every command should make its flag requirements and fields explicit (aside from the few true globals)
3. Ensuring our kingpin configuration did not enforce any new and unergonomic flag ordering constraints. kingpin flags don't have nice inheritance rules so defining common arguments on the `tbot start` level led to:
   * ugly positioning requirements: `tbot start --proxy-server foo.com identity --destination ./bar`
   * inability to properly exclude old/renamed arguments from new commands

To attempt to solve everything at once (very dangerous 😬) I've attempted to adapt `tctl`'s [`TryRun()`](https://github.com/gravitational/teleport/blob/master/tool/tctl/common/bots_command.go#L134) to allow for reusable commands that we can execute as either `start` or `configure` subcommands. Defining new "modern" style subcommands is quite lightweight as the shared set of parameters can be embedded easily, and the "start" and "configure" actions can be passed in at init time.

changelog: Introduced use-case specific commands to tbot to provide a better out-of-box experience.

timothyb89 committed:
Machine ID: Refactor `tbot` CLI

This significantly refactors the tbot CLI. In addition to a major
overhaul of our CLI config handling, it also exposes a number of new
subcommands for starting more than just an identity output via pure
CLI.

timothyb89 commented on lib/tbot/config/cli.go:
```diff
@@ -0,0 +1,579 @@
+		strings.Join(SupportedJoinMethods, ", "),
+	)
+
+	cmd.Flag("proxy-server", "Address of the Teleport Proxy Server.").Envar(proxyServerEnvVar).StringVar(&args.ProxyServer)
```
It occurs to me --auth-server is not actually deprecated, just no longer preferred, and will need to be supported here. Will re-add.

timothyb89 commented on lib/tbot/config/config.go:
```diff
@@ -677,6 +681,7 @@ func destinationFromURI(uriString string) (bot.Destination, error) {
 // FromCLIConf loads bot config from CLI parameters, potentially loading and
 // merging a configuration file if specified. CheckAndSetDefaults() will
 // be called. Note that CLI flags, if specified, will override file values.
+// TODO: Remove this function.
```
(I expect to remove `FromCLIConf` outright, and significantly prune the `CLIConf` struct)

timothyb89 replied:
This struct is now removed in favor of command-specific structs (and a small dedicated GlobalArgs for the few fields where globals actually make sense)

timothyb89 commented on lib/tbot/config/cli.go:
```diff
@@ -0,0 +1,579 @@
+	"github.com/gravitational/trace"
+)
+
+type globalArgs struct {
```
I've still not decided on a strategy for these global args, so this is dead code at the moment. I'll sort out some solution here if we decide to greenlight this refactor as a whole.

timothyb89 replied:
The global arg situation is resolved, I think. I found a bunch of incorrect usages throughout our CLI but they're all now standardized in a GlobalArgs struct that gets uniformly applied to the config end result.

strideynet commented on lib/tbot/config/cli.go:
```diff
@@ -0,0 +1,579 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package config
```
My only "concern" here is pushing this stuff down into the already fairly over-scoped "config" package. We should keep in mind that `tbot` is seeing embedded usage now, and we're going to be inflicting any dependencies introduced into `lib/tbot` and `lib/tbot/config` on any consumer. If it's possible to have this as a separate `lib/tbot/cli` without a large amount of work, that would be nice, but otherwise we can revisit this.

timothyb89 replied:
Agreed, I've moved it into its own package. And to fully commit to separating `lib/tbot/config` from CLI stuff, all kingpin related functionality is now removed from that package and the import is gone.

timothyb89 committed:
Move new CLI handling into lib/tbot/cli; add all new subcommands

This moves the new CLI handling code into `lib/tbot/cli`, splits
`cli.go` into several better organized files, and adds all missing
subcommands for major tbot functionality.

timothyb89 committed:
Add sane global handling, convert most other commands to new style

This introduces a dedicated global args struct and removes CLIConf.
As most other non-start (and configure) commands depended on that
effectively global namespace, this refactors all of them to use the
new style with properly namespaced command structs.

This also introduces a `genericExecutorHandler` helper to simplify
subcommands that run an arbitrary action without modifying the
config.

This also solves a number of longstanding CLI handling bugs resulting
from the shared arg namespace. A few args were moved to globals that
should always have been exposed, so now things like `--log-format`
will work consistently.

A few commands were left unconverted since they don't make use of
any of the global args or config loading machinery. They could be
converted easily but are technically simpler as written. We may
opt to convert them, or to do so later.

Some headway on testing, but more work is needed.

timothyb89 committed:
First batch of tests

This adds the first batch of tests on the new CLI along with a bunch
of associated changes made while fixing these tests and others.

This also splits off 2 new flag helpers: LegacyDestinationDirArgs and
AuthProxyArgs. These are reusable embeddable structs to handle
legacy-style --destination-dir behavior, and modern --auth-server
and --proxy-server. `tbot init` and its tests had a subtle dependency
old-style `--destination-dir` handling so we can now reuse that.

There may be other subcommands that require something similar.

timothyb89 committed:
Remove outdated TODO

timothyb89 committed:
Restore TestConfigCLIOnlySample

timothyb89 committed:
Fix lints

timothyb89 commented on lib/tbot/cli/db.go:
```diff
@@ -0,0 +1,68 @@
+	Cluster        string
+	ProxyServer    string
+
+	// LegacyProxy is the legacy --proxy flag.
+	// TODO(timothyb89): DELETE IN 17.0.0
+	// TODO(timothyb89): Or maybe remove in this PR.
+	LegacyProxyFlag string
```
@strideynet do we just want to remove this flag now? It's ported over as is, but I figured it was worth discussing. Happy to follow up with a removal later.

strideynet replied:
Let's leave it in for now - probably easier to remove in an atomic PR.

timothyb89 commented on lib/tbot/cli/globals.go:
```diff
@@ -0,0 +1,114 @@
+		cfg.Debug = g.Debug
+	}
+
+	// TODO: in previous versions, `insecure` is handled _after_
+	// BotConfig.CheckAndSetDefaults(). This flag is checked and setting it here
+	// *will* cause a behavioral change, so make sure the new behavior is sane.
+	// (It is unclear why this was done.)
+	if g.Insecure {
+		cfg.Insecure = true
+	}
```
I'm a bit inclined to believe this old behavior was a bug and would allow `--ca-pins` and `--ca-path` to be set alongside `--insecure` which we try to consider an error in `BotConfig.CheckAndSetDefaults()`.

strideynet replied:
It definitely seems like a case of user confusion if someone is specifying these flags together and I'd definitely rather explicitly block it. In my experience, gracefully handling invalid configurations only leads to more user confusion as they start to build up a mental model.

timothyb89 replied:
👍 I'll keep the behavior as written here and remove the TODO. I figure this slight behavior change is one more vote in the "don't backport to v16, at least not yet" bucket.

timothyb89 committed:
Fix missing embedded flag init

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/tbot-cli-overhaul

timothyb89 commented:
I think I'm ready to open this for review. I still want to spend tomorrow adding unit tests to verify each command's individual flag handling, but also want to start collecting feedback before I'm OoO this week.

strideynet commented on tool/tbot/main.go:
So nice seeing the hundreds of flags pulled out of here.

strideynet commented on tool/tbot/main.go:
```diff
@@ -66,126 +60,92 @@ access Teleport protected resources in the same way your engineers do!
+		cli.NewSPIFFEX509SVIDCommand(configureCmd, buildConfigAndConfigure(ctx, globalCfg, &configureOutPath, stdout)),
+	)
+
+	// TODO: workload id / spiffe service subcommand?
```
Let's leave this out for now I think - the framework you've added here makes it easy enough for me to add this once I've identified the simplest possible use-case.

strideynet commented on lib/tbot/cli/start_spiffe_x509_svid.go:
```diff
@@ -0,0 +1,86 @@
+	cmd.Flag("include-federated-trust-bundles", "If set, include federated trust bundles in the output").BoolVar(&c.IncludeFederatedTrustBundles)
+	cmd.Flag("svid-path", "A SPIFFE ID to request, prefixed with '/'").Required().StringVar(&c.SVIDPath)
+	cmd.Flag("svid-hint", "An optional hint for consumers of the SVID to aid in identification").StringVar(&c.SVIDHint)
+	cmd.Flag("request-dns-san", "A DNS name that should be included in the SVID. Repeatable.").StringsVar(&c.DNSSANs)
```
```suggestion
	cmd.Flag("dns-san", "A DNS name that should be included in the SVID. Repeatable.").StringsVar(&c.DNSSANs)
```

strideynet commented on lib/tbot/cli/start_spiffe_x509_svid.go:
```diff
@@ -0,0 +1,86 @@
+	cmd.Flag("svid-path", "A SPIFFE ID to request, prefixed with '/'").Required().StringVar(&c.SVIDPath)
+	cmd.Flag("svid-hint", "An optional hint for consumers of the SVID to aid in identification").StringVar(&c.SVIDHint)
+	cmd.Flag("request-dns-san", "A DNS name that should be included in the SVID. Repeatable.").StringsVar(&c.DNSSANs)
+	cmd.Flag("request-ip-san", "An IP address that should be included in the SVID. Repeatable.").StringsVar(&c.IPSANs)
```
```suggestion
	cmd.Flag("ip-san", "An IP address that should be included in the SVID. Repeatable.").StringsVar(&c.IPSANs)
```

strideynet reviewed (approved):
This is looking really good with a separate package for the CLI concerns. This has definitely come together nicely.

I think at a later date, I'd be slightly curious about moving the actual handler code down into `cli` or tying it more to the command constructors themselves. But I think that's definitely out of scope for this PR and would realistically bring only minor improvements.

Approving this on the basis of some more tests :')

strideynet commented:
Let's make the changelog message a little more user facing as well. Perhaps:

changelog: Introduced use-case specific commands to `tbot` to provide a better out-of-box experience.

timothyb89 committed:
Code review feedback

timothyb89 committed:
Add docstrings

timothyb89 committed:
Add unit tests for all "start" commands and globals

Also includes a few minor fixes discovered in testing - DiagAddr
was not handled properly in shared args, and removed an unused field.

timothyb89 committed:
Fix imports

strideynet committed:
Add helper for repetitive test

strideynet committed:
Fix command name

strideynet committed:
Adjust behaviour of `tbot kube credentials` command

strideynet committed:
Add tests for remaining commands

strideynet committed:
Appease linter

strideynet commented on lib/tbot/cli/cli.go:
```diff
@@ -0,0 +1,225 @@
+// genericExecutorHandler is a helper that can be embedded to provide a simpler
+// TryRun implementation that just runs a function. These functions can be
+// passed in while building the CLI to more easily glue behaviors together, or
+// specified directly.
+type genericExecutorHandler[T any] struct {
+	cmd  *kingpin.CmdClause
+	args *T
+
+	// actions is a list of functions to run when `TryRun` matches the `cmd`.
+	// Generally at most one action should be exposed to the top level glue in
+	// main, but commands might want to inject some handler logic for e.g.
+	// flag migrations.
+	actions []func(*T) error
+}
+
+// newGenericExecutorHandler creates a genericExecutorHandler with the given
+// command and action to execute when that command is matched.
+func newGenericExecutorHandler[T any](cmd *kingpin.CmdClause, args *T, actions ...func(*T) error) *genericExecutorHandler[T] {
+	return &genericExecutorHandler[T]{
+		cmd:     cmd,
+		args:    args,
+		actions: actions,
+	}
+}
+
+func (e *genericExecutorHandler[T]) TryRun(cmd string) (match bool, err error) {
+	switch cmd {
+	case e.cmd.FullCommand():
+		for _, action := range e.actions {
+			err = action(e.args)
+			if err != nil {
+				break
+			}
+		}
+	default:
+		return false, nil
+	}
+
+	return true, trace.Wrap(err)
+}
```
TimB's on leave until next week - I figure that since this PR is already pretty large - the overall idea of extracting this to something reusable makes sense but we perhaps ought to do that in another PR. I think there's a few finishing touches/refinements we'll want to add if other folks are going to use this.

strideynet committed:
Add godocs
